File: common/src/main/java/dev/gallon/motorassistance/common/domain/Timer.java
package dev.gallon.motorassistance.common.domain;

public class Timer {
    private long startMs;
    private boolean stopped;

    public Timer() {
        startMs = getSystemMilliseconds();
        stopped = true;
    }

    public void start() {
        startMs = getSystemMilliseconds();
        stopped = false;
    }

    public void stop() {
        stopped = true;
    }

    public boolean timeElapsed(long milliseconds) {
        return !stopped && timeElapsed() >= milliseconds;
    }

    long timeElapsed() {
        return getSystemMilliseconds() - startMs;
    }

    public boolean stopped() {
        return stopped;
    }

    private long getSystemMilliseconds() {
        return System.nanoTime() / 1000000L;
    }
}

File: common/src/main/kotlin/dev/gallon/motorassistance/common/domain/Timer.kt
package dev.gallon.motorassistance.common.domain

class Timer {

    private var startMs: Long
    private var stopped: Boolean

    init {
        startMs = getSystemMilliseconds()
        stopped = true
    }

    fun start() {
        startMs = getSystemMilliseconds()
        stopped = false
    }

    fun stop() {
        stopped = true
    }

    fun timeElapsed(milliseconds: Long): Boolean = !stopped && timeElapsed() >= milliseconds
    fun timeElapsed(): Long = getSystemMilliseconds() - startMs
    fun stopped(): Boolean = stopped

    private fun getSystemMilliseconds() = System.nanoTime() / 1000000L
}
